torchvision/prototype/datasets/_builtin/caltech.py

```diff
@@ -0,0 +1,209 @@
+            anns_dp,
+            key_fn=self._images_key_fn,
+            ref_key_fn=self._anns_key_fn,
+            buffer_size=INFINITE_BUFFER_SIZE,
+            keep_key=True,
+        )
+        return Mapper(dp, self._collate_and_decode_sample, fn_kwargs=dict(decoder=decoder))
+
+    def generate_categories_file(self, root: Union[str, pathlib.Path]) -> None:
+        dp = self.resources(self.default_config)[0].to_datapipe(pathlib.Path(root) / self.name)
+        dp = TarArchiveReader(dp)
+        dp = Filter(dp, self._is_not_background_image)
+        dir_names = {pathlib.Path(path).parent.name for path, _ in dp}
+        create_categories_file(HERE, self.name, sorted(dir_names))
+
+
+class Caltech256(Dataset):
+    @property
+    def info(self) -> DatasetInfo:
+        return DatasetInfo(
+            "caltech256",
+            categories=HERE / "caltech256.categories",
+            homepage="http://www.vision.caltech.edu/Image_Datasets/Caltech256",
+        )
+
+    def resources(self, config: DatasetConfig) -> List[OnlineResource]:
+        return [
+            HttpResource(
+                "http://www.vision.caltech.edu/Image_Datasets/Caltech256/256_ObjectCategories.tar",
+                sha256="08ff01b03c65566014ae88eb0490dbe4419fc7ac4de726ee1163e39fd809543e",
+            )
+        ]
+
+    def _is_not_rogue_file(self, data: Tuple[str, Any]) -> bool:
+        path = pathlib.Path(data[0])
+        return path.name != "RENAME2"
+
+    def _collate_and_decode_sample(
+        self,
+        data: Tuple[str, io.IOBase],
+        *,
+        decoder: Optional[Callable[[io.IOBase], torch.Tensor]],
+    ) -> Dict[str, Any]:
+        path, buffer = data
+
+        dir_name = pathlib.Path(path).parent.name
+        label_str, category = dir_name.split(".")
+        label = torch.tensor(int(label_str))
+
+        return dict(label=label, category=category, image=decoder(buffer) if decoder else buffer)
+
+    def _make_datapipe(
+        self,
+        resource_dps: List[IterDataPipe],
+        *,
+        config: DatasetConfig,
+        decoder: Optional[Callable[[io.IOBase], torch.Tensor]],
+    ) -> IterDataPipe[Dict[str, Any]]:
+        dp = resource_dps[0]
+        dp = TarArchiveReader(dp)
+        dp = Filter(dp, self._is_not_rogue_file)
+        # FIXME: add this after https://github.com/pytorch/pytorch/issues/65808 is resolved
+        # dp = Shuffler(dp, buffer_size=INFINITE_BUFFER_SIZE)
+        return Mapper(dp, self._collate_and_decode_sample, fn_kwargs=dict(decoder=decoder))
+
+    def generate_categories_file(self, root: Union[str, pathlib.Path]) -> None:
+        dp = self.resources(self.default_config)[0].to_datapipe(pathlib.Path(root) / self.name)
+        dp = TarArchiveReader(dp)
+        dir_names = {pathlib.Path(path).parent.name for path, _ in dp}
+        categories = [name.split(".")[1] for name in sorted(dir_names)]
+        create_categories_file(HERE, self.name, categories)
+
+
+if __name__ == "__main__":
```

Comment:
Question: do we want to keep this for all datasets?

Reply:
The idea behind this is that we should keep track of how we generated the category files. With this it is easy to regenerate if we need to. Just call `python -m torchvision.prototype.datasets._builtin.caltech`. Maybe we could also do an aggregation to be able to generate "everything", but I would keep that for the future.